desc/integrals/basis.py

```diff
@@ -58,6 +85,627 @@ def _in_epigraph_and(is_intersect, df_dy_sign, /):
+            Grid of (x, y) points for optimal interpolation.
+
+        """
+        x = fourier_pts(M)
+        y = cheb_pts(N, domain, lobatto)
+        if L is None:
+            coords = (x, y)
+        else:
+            if isposint(L):
+                L = jnp.flipud(jnp.linspace(1, 0, L, endpoint=False))
+            coords = (jnp.atleast_1d(L), x, y)
+        coords = tuple(map(jnp.ravel, jnp.meshgrid(*coords, indexing="ij")))
+        return jnp.column_stack(coords)
+
+    def evaluate(self, M, N):
+        """Evaluate Fourier-Chebyshev series.
+
+        Parameters
+        ----------
+        M : int
+            Grid resolution in x direction. Preferably power of 2.
+        N : int
+            Grid resolution in y direction. Preferably power of 2.
+
+        Returns
+        -------
+        fq : jnp.ndarray
+            Shape (..., M, N)
+            Fourier-Chebyshev series evaluated at
+            ``FourierChebyshevSeries.nodes(M,N,L,self.domain,self.lobatto)``.
+
+        """
+        fq = idct(
+            irfft(self._c, n=M, axis=-2, norm="forward"),
+            type=2 - self.lobatto,
+            n=N,
+            axis=-1,
+        ) * (N - self.lobatto)
+        return fq
+
+    def harmonics(self):
+        """Spectral coefficients aₘₙ of the interpolating trigonometric polynomial.
+
+        Transform Fourier interpolant harmonics to Nyquist trigonometric
+        interpolant harmonics so that the coefficients are all real.
+
+        Returns
+        -------
+        a_mn : jnp.ndarray
+            Shape (..., M, N).
+            Real valued spectral coefficients for Fourier-Chebyshev series.
+
+        """
+        a_mn = harmonic(cheb_from_dct(self._c), self.M, axis=-2)
+        assert a_mn.shape[-2:] == (self.M, self.N)
+        return a_mn
+
+    def compute_cheb(self, x):
+        """Evaluate Fourier series at ``x`` to obtain set of 1D Chebyshev coefficients.
+
+        Parameters
+        ----------
+        x : jnp.ndarray
+            Points to evaluate Fourier series.
+
+        Returns
+        -------
+        cheb : PiecewiseChebyshevSeries
+            Chebyshev coefficients αₙ(x=``x``) for f(x, y) = ∑ₙ₌₀ᴺ⁻¹ αₙ(x) Tₙ(y).
+
+        """
+        # Add axis to broadcast against Chebyshev coefficients.
+        x = jnp.atleast_1d(x)[..., jnp.newaxis]
+        # Add axis to broadcast against multiple x values.
+        cheb = cheb_from_dct(
+            irfft_non_uniform(x, self._c[..., jnp.newaxis, :, :], self.M, axis=-2)
+        )
+        assert cheb.shape[-2:] == (x.shape[-2], self.N)
+        return PiecewiseChebyshevSeries(cheb, self.domain)
+
+
+class PiecewiseChebyshevSeries(IOAble):
+    """Chebyshev series.
+
+    { fₓ | fₓ : y ↦ ∑ₙ₌₀ᴺ⁻¹ aₙ(x) Tₙ(y) }
+    and Tₙ are Chebyshev polynomials on [−yₘᵢₙ, yₘₐₓ]
+
+    Attributes
+    ----------
+    cheb : jnp.ndarray
+        Shape (..., M, N).
+        Chebyshev coefficients αₙ(x) for fₓ(y) = ∑ₙ₌₀ᴺ⁻¹ αₙ(x) Tₙ(y).
+    M : int
```

Comment:
these are already declared as `@property` so you shouldn't need to redocument them here